chain-evm/src/state/trie.rs

```diff
@@ -0,0 +1,78 @@
+use imhamt::{Hamt, RemoveError};
+
+use std::collections::hash_map::DefaultHasher;
+use std::convert::Infallible;
+use std::hash::Hash;
+
+/// An immutable structure to represent any of EVM tries.
+#[derive(Clone, Default)]
+pub struct Trie<K: Clone + Hash + Eq, V: Clone>(Hamt<DefaultHasher, K, V>);
+
+impl<K: Clone + Hash + Eq, V: Clone> Trie<K, V> {
+    pub fn new() -> Self {
+        Self(Hamt::new())
+    }
+
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.0.lookup(key)
+    }
+
+    /// Check if this trie contains a given key.
+    pub fn contains(&self, key: &K) -> bool {
+        self.0.contains_key(key)
+    }
+
+    /// Put a value into a trie replacing an exisisting value if there was any.
+    pub fn put(self, key: K, value: V) -> Self {
+        // using two branches instead of `Hamt::insert_or_update` to avoid unnecessary cloning
+        let new_state = if self.0.contains_key(&key) {
+            self.0
+                .update(&key, |_| Ok::<_, Infallible>(Some(value)))
+                .expect("we already checked that the key is present")
+        } else {
+            self.0
+                .insert(key, value)
+                .expect("we already checked that the key does not exist")
+        };
```

Comment:
Could use `.insert_or_update`.

Reply:
I left a comment in the beginning of this function as per why I did not use it. It would result in a code like `self.0.insert_or_update(key, value.clone(), |_| Some(value))`. Cloning is completely unnecessary here and should be avoided IMO.

Reply:
Hmm, there should be some Entry-like API to Hamt to deal with this.